### PR TITLE
Joule lenz law

### PR DESCRIPTION
**Pull request by cslade1970**

### CODE REVIEW

File: examples/electricity/amount_energy_from_voltage_time_resistance.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+SI.set_quantity_scale_factor(P_example, 1500 * units.watt)
+m_example = units.Quantity('m_example')
+SI.set_quantity_dimension(m_example, units.mass)
+SI.set_quantity_scale_factor(m_example, 0.5 * units.kilogram)
+C_example = units.Quantity('C_example')
+SI.set_quantity_dimension(C_example, units.energy / (units.mass * units.temperature))
+SI.set_quantity_scale_factor(C_example, 4200 * units.joule / (units.kilogram * units.kelvin))
+t1_example = units.Quantity('t1_example')
+SI.set_quantity_dimension(t1_example, units.temperature)
+SI.set_quantity_scale_factor(t1_example, 373 * units.kelvin)
```
Comment (blackyblack) [inline examples/electricity/amount_energy_from_voltage_time_resistance.py:26]: What are these numbers? Check if it possible to convert from Celsius degrees to Kelvin or add an explanation what you are doing here.

File: examples/electricity/amount_energy_from_voltage_time_resistance.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+t2_example = units.Quantity('t2_example')
+SI.set_quantity_dimension(t2_example, units.temperature)
+SI.set_quantity_scale_factor(t2_example, 308 * units.kelvin)
+P, t, m , C, t1, t2 = symbols('P t m C t2 t1')
+Q1 = joule_lenz_law.law.subs({
+    joule_lenz_law.operating_voltage / joule_lenz_law.resistance_heater: ohm_law.current,
+    ohm_law.current * joule_lenz_law.operating_voltage: P, joule_lenz_law.operating_time: t
```
Comment (blackyblack) [inline examples/electricity/amount_energy_from_voltage_time_resistance.py:33]: You should not replace values directly. Apply law to determine how voltage, resistance and current are tied together.

File: examples/electricity/amount_energy_from_voltage_time_resistance.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+})
+Q2 = operate_energy.law.subs({operate_energy.body_mass: m, operate_energy.specific_heat_capacity: C,
+     operate_energy.final_temperature: t2, operate_energy.initial_temperature: t1
```
Comment (blackyblack) [inline examples/electricity/amount_energy_from_voltage_time_resistance.py:36]: There is no much sense in substituting unknown variables with unknown variables. Simply use variables directly from the law module.

File: examples/electricity/amount_energy_from_voltage_time_resistance.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+})
+law = Eq(Q1, Q2)
+solved = solve(law, t, dict=True)[0][t]
+answer = Eq(t, solved)
+print("\nFormula is:\n\n {}".format(pretty(answer, use_unicode=False)))
+law_example = Eq(Q1, Q2).subs({
```
Comment (blackyblack) [inline examples/electricity/amount_energy_from_voltage_time_resistance.py:42]: Eq(Q1, Q2) is assigned to `law`. Use this variable.

File: examples/electricity/amount_energy_from_voltage_time_resistance.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    P: convert_to(P_example, units.watt).subs(units.watt, 1).evalf(5),
+    m: convert_to(m_example, units.kilogram).subs(units.kilogram, 1).evalf(4),
+    C: convert_to(C_example, units.joule / (units.kilogram * units.kelvin)).subs(units.joule / (units.kilogram * units.kelvin), 1).evalf(6),
+    t2: convert_to(t1_example, units.kelvin).subs(units.kelvin, 1).evalf(3),
+    t1: convert_to(t2_example, units.kelvin).subs(units.kelvin, 1).evalf(3)
```
Comment (blackyblack) [inline examples/electricity/amount_energy_from_voltage_time_resistance.py:47]: You should not convert input parameters. Convert output instead.

File: examples/electricity/amount_energy_from_voltage_time_resistance.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+})
+solved_example = solve(law_example, t, dict=True)[0][t]
+print("\n For P = {} {}, water mass = {} {}, t2 = {} {}, t1 = {} {}, cooling time = {} {}\n"
+    .format(
+        convert_to(P_example, units.watt).subs(units.watt, 1).evalf(5),
+        units.watt,
+        convert_to(m_example, units.kilogram).subs(units.kilogram, 1).evalf(4),
+        units.kilogram,
+        convert_to(t1_example, units.kelvin).subs(units.kelvin, 1).evalf(3),
+        units.kelvin,
```
Comment (blackyblack) [inline examples/electricity/amount_energy_from_voltage_time_resistance.py:57]: Use Celsius degrees for human readable output.

File: symplyphysics/definitions/amount_energy_from_mass_and_temperature.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+from symplyphysics import (
+    symbols, Eq, pretty, solve, Quantity, units,
+    validate_input, validate_output, expr_to_quantity
+)
+
+# Description
+# Amount of energy for body heat Q = C * m * (t2 - t1)
+# where:
+# C - specific heat capacity
```
Comment (blackyblack) [inline symplyphysics/definitions/amount_energy_from_mass_and_temperature.py:9]: Add high level description of this parameter. Add wiki link if it cannot be easily explained.

File: symplyphysics/definitions/amount_energy_from_mass_and_temperature.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+law = Eq(amount_energy, specific_heat_capacity * body_mass * (final_temperature - initial_temperature))
+
+def print():
+    return pretty(law, use_unicode=False)
+
+@validate_input(specific_heat_capacity_=units.energy / (units.mass * units.temperature), body_mass_=units.mass,
+    final_temperature_=units.temperature, initial_temperature_=units.temperature
+)
+@validate_output(units.energy)
+def calculate_amount_energy(specific_heat_capacity_: Quantity, body_mass_: Quantity, final_temperature_: Quantity,
```
Comment (blackyblack) [inline symplyphysics/definitions/amount_energy_from_mass_and_temperature.py:25]: Initial value in parameters usually comes first.

File: symplyphysics/definitions/power_is_proportional_voltage_and_current.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+from symplyphysics import (
+    symbols, Eq, pretty, solve, Quantity, units,
+    validate_input, validate_output, expr_to_quantity
+)
+
+# Description
+# Power of current is proportional current and voltage
```
Comment (blackyblack) [inline symplyphysics/definitions/power_is_proportional_voltage_and_current.py:7]:
```suggestion
# Power of current is proportional to current and voltage
```

File: symplyphysics/laws/electricity/amount_energy_from_voltage_time_resistance.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+from symplyphysics import (
+    symbols, Eq, pretty, solve, Quantity, units,
+    validate_input, validate_output, expr_to_quantity,
+)
+# Description
+# Amount heat energy of current Q = U^2 *t / R
```
Comment (blackyblack) [inline symplyphysics/laws/electricity/amount_energy_from_voltage_time_resistance.py:6]:
```suggestion
# Amount heat energy of current Q = U**2 * t / R
```

File: symplyphysics/laws/electricity/amount_energy_from_voltage_time_resistance.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+# where:
+# U - voltage
+# t - time
+# R - resistance
+amount_energy, operating_voltage, operating_time, resistance_heater = symbols(
+    'amount_energy, operating_voltage, operating_time, resistance_heater')
```
Comment (blackyblack) [inline symplyphysics/laws/electricity/amount_energy_from_voltage_time_resistance.py:12]: Remove commas for symbols.

File: test/definitions/amount_energy_from_mass_and_temperature_test.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    C = units.Quantity('C')
+    SI.set_quantity_dimension(C, units.energy / (units.mass * units.temperature))
+    SI.set_quantity_scale_factor(C, 4200 * units.joule / (units.kilogram * units.kelvin))
+    m = units.Quantity('m')
+    SI.set_quantity_dimension(m, units.mass)
+    SI.set_quantity_scale_factor(m, 0.5 * units.kilogram)
+    t2 = units.Quantity('t2')
+    SI.set_quantity_dimension(t2, units.temperature)
+    SI.set_quantity_scale_factor(t2, 373 * units.kelvin)
+    t1 = units.Quantity('t1')
```
Comment (blackyblack) [inline test/definitions/amount_energy_from_mass_and_temperature_test.py:22]: Define t1 first.

File: test/definitions/amount_energy_from_mass_and_temperature_test.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    SI.set_quantity_dimension(t1, units.temperature)
+    SI.set_quantity_scale_factor(t1, 273 * units.kelvin)
+    Args = namedtuple('Args', ['C', 'm', 't2', 't1'])
+    return Args(C=C, m=m, t2=t2, t1=t1)
+
+def test_basic_amount(test_args):
+    result = amount_energy.calculate_amount_energy(test_args.C, test_args.m, test_args.t2, test_args.t1)
+    assert SI.get_dimension_system().equivalent_dims(result.dimension, units.energy)
+    result_energy = convert_to(result, units.joule).subs(units.joule, 1).evalf(7)
+    assert result_energy == approx(210000.0, 0.0005)
```
Comment (blackyblack) [inline test/definitions/amount_energy_from_mass_and_temperature_test.py:32]: How come you heat to 100 degrees but test pass? Looks like made up numbers.

File: test/definitions/amount_energy_from_mass_and_temperature_test.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+def test_bad_final_temperature(test_args):
+    bt2 = units.Quantity('bt2')
+    SI.set_quantity_dimension(bt2, units.mass)
+    SI.set_quantity_scale_factor(bt2, 1 * units.kilogram)
+    with raises(errors.UnitsError):
+        amount_energy.calculate_amount_energy(test_args.C, test_args.m, bt2, test_args.t1)
+    with raises(TypeError):
+        amount_energy.calculate_amount_energy(test_args.C, test_args.m, 100, test_args.t1)
+
+def test_bad_initial_temperature(test_args):
```
Comment (blackyblack) [inline test/definitions/amount_energy_from_mass_and_temperature_test.py:62]: Combine tests for bad temperature together.

File: test/definitions/power_is_proportional_voltage_and_current_test.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
+from collections import namedtuple
+from pytest import approx, fixture, raises
+
+from symplyphysics import (
+    units, convert_to, SI, errors
+)
+from symplyphysics.definitions import power_is_proportional_voltage_and_current as power_def
+# What is the capacity of the battery at 9 V and a current of 0.5 A?
```
Comment (blackyblack) [inline test/definitions/power_is_proportional_voltage_and_current_test.py:8]: It is not about capacity.

File: test/electricity/amount_energy_from_voltage_time_resistance_test.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
+from collections import namedtuple
+from pytest import approx, fixture, raises
+
+from symplyphysics import (
+    units, convert_to, SI, errors
+)
+from symplyphysics.laws.electricity import amount_energy_from_voltage_time_resistance as joule_lenz_law
+
+# How much energy does it take to heat 0.5 kilograms of water to 50 kelvin?
+# Specific heat capacity of water is 4200 J/kg*K , ignore losses.
```
Comment (blackyblack) [inline test/electricity/amount_energy_from_voltage_time_resistance_test.py:10]: Wrong test description.

File: symplyphysics/definitions/amount_energy_from_mass_and_temperature.py [hunk 17]
```diff
@@ -10,21 +10,23 @@
...
 # t1 - initial temperature
-amount_energy, specific_heat_capacity, body_mass, final_temperature, initial_temperature = symbols(
-    'amount_energy specific_heat_capacity body_mass final_temperature initial_temperature'
+# Note: specific heat capacity C is parameter indicates how much energy in joules
+# must be expended to heat a unit mass of a substance by one unit temperature.
+amount_energy, specific_heat_capacity, body_mass, temperature_begin, temperature_end = symbols(
+    'amount_energy specific_heat_capacity body_mass temperature_begin temperature_end'
 )
-law = Eq(amount_energy, specific_heat_capacity * body_mass * (final_temperature - initial_temperature))
+law = Eq(amount_energy, specific_heat_capacity * body_mass * (temperature_begin - temperature_end))
```
Comment (blackyblack) [inline symplyphysics/definitions/amount_energy_from_mass_and_temperature.py:18]: Does not correspond to description.

File: symplyphysics/definitions/amount_energy_from_mass_and_temperature.py [hunk 18]
```diff
@@ -10,21 +10,23 @@
 # m - body mass
 # t2 - final temperature
 # t1 - initial temperature
-amount_energy, specific_heat_capacity, body_mass, final_temperature, initial_temperature = symbols(
-    'amount_energy specific_heat_capacity body_mass final_temperature initial_temperature'
+# Note: specific heat capacity C is parameter indicates how much energy in joules
+# must be expended to heat a unit mass of a substance by one unit temperature.
```
Comment (blackyblack) [inline symplyphysics/definitions/amount_energy_from_mass_and_temperature.py:14]: Also add a note about resulting amount of energy. Is it the generated energy or consumed energy?

File: symplyphysics/laws/electricity/amount_energy_from_voltage_time_resistance.py [hunk 19]
```diff
@@ -3,13 +3,13 @@
     validate_input, validate_output, expr_to_quantity,
 )
 # Description
-# Amount heat energy of current Q = U^2 *t / R
+# Amount heat energy of current Q = U**2 *t / R
```
Comment (blackyblack) [inline symplyphysics/laws/electricity/amount_energy_from_voltage_time_resistance.py:6]: Add a note if it is generated energy or consumed.

File: test/definitions/amount_energy_from_mass_and_temperature_test.py [hunk 20]
```diff
@@ -16,54 +16,52 @@ def test_args():
...
-    SI.set_quantity_dimension(t2, units.temperature)
-    SI.set_quantity_scale_factor(t2, 373 * units.kelvin)
     t1 = units.Quantity('t1')
     SI.set_quantity_dimension(t1, units.temperature)
-    SI.set_quantity_scale_factor(t1, 273 * units.kelvin)
-    Args = namedtuple('Args', ['C', 'm', 't2', 't1'])
+    SI.set_quantity_scale_factor(t1, 323 * units.kelvin)
+    t2 = units.Quantity('t2')
+    SI.set_quantity_dimension(t2, units.temperature)
+    SI.set_quantity_scale_factor(t2, 273 * units.kelvin)
```
Comment (blackyblack) [inline test/definitions/amount_energy_from_mass_and_temperature_test.py:24]: Does not correspond to description. You are supposed to heat the body.

File: test/definitions/power_is_proportional_voltage_and_current_test.py [hunk 21]
```diff
@@ -5,7 +5,7 @@
     units, convert_to, SI, errors
 )
 from symplyphysics.definitions import power_is_proportional_voltage_and_current as power_def
-# What is the capacity of the battery at 9 V and a current of 0.5 A?
+# What is the power can release the battery at 9 V and a current of 0.5 A?
```
Comment (blackyblack) [inline test/definitions/power_is_proportional_voltage_and_current_test.py:8]:
```suggestion
# What power can be released by the battery at 9 V and a current of 0.5 A?
```

File: test/electricity/amount_energy_from_voltage_time_resistance_test.py [hunk 22]
```diff
@@ -6,16 +6,16 @@
 )
 from symplyphysics.laws.electricity import amount_energy_from_voltage_time_resistance as joule_lenz_law
 
-# How much energy does it take to heat 0.5 kilograms of water to 50 kelvin?
-# Specific heat capacity of water is 4200 J/kg*K , ignore losses.
+#How much energy can a household electric kettle produce to heat water in one minute
+#at 220 volts and a heater resistance of 36 ohms?
```
Comment (blackyblack) [inline test/electricity/amount_energy_from_voltage_time_resistance_test.py:10]: Add space after #

File: test/definitions/amount_energy_from_mass_and_temperature_test.py (hunk 20 above)
Comment (cslade1970) [reply test/definitions/amount_energy_from_mass_and_temperature_test.py:24]: It's not clear what the discrepancy is, added the conversions.

Comment (blackyblack) [reply test/definitions/amount_energy_from_mass_and_temperature_test.py:24]: T1 is supposed to be initial temperature.

Comment (cslade1970) [reply test/definitions/amount_energy_from_mass_and_temperature_test.py]: corrected the parameter transfer order

File: test/definitions/amount_energy_from_mass_and_temperature_test.py [hunk 23]
```diff
@@ -19,15 +20,15 @@ def test_args():
...
     t2 = units.Quantity('t2')
     SI.set_quantity_dimension(t2, units.temperature)
-    SI.set_quantity_scale_factor(t2, (273 - 273) * units.kelvin)
+    SI.set_quantity_scale_factor(t2, (323 - CELSIUS_TO_KELVIN_OFFSET) * units.kelvin)
     Args = namedtuple('Args', ['C', 'm', 't1', 't2'])
     return Args(C=C, m=m, t2=t2, t1=t1)
 
 def test_basic_amount(test_args):
-    result = amount_energy.calculate_amount_energy(test_args.C, test_args.m, test_args.t1, test_args.t2)
+    result = amount_energy.calculate_amount_energy(test_args.C, test_args.m, test_args.t2, test_args.t1)
```
Comment (blackyblack) [inline test/definitions/amount_energy_from_mass_and_temperature_test.py:31]: Why did you change order of parameters? T1 is supposed to be initial temperature and should come first.

File: symplyphysics/definitions/amount_energy_from_mass_and_temperature.py [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+    validate_input, validate_output, expr_to_quantity
+)
+
+# Description
+# Amount of energy for body heat Q = C * m * (t2 - t1)
+# where:
+# C - specific heat capacity
+# m - body mass
+# t1 - temperature_origin (origin process)
+# t2 - temperature_end ( end process)
```
Comment (blackyblack) [inline symplyphysics/definitions/amount_energy_from_mass_and_temperature.py:12]:
```suggestion
# t2 - final temperature
```

File: symplyphysics/definitions/amount_energy_from_mass_and_temperature.py [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+    symbols, Eq, pretty, solve, Quantity, units,
+    validate_input, validate_output, expr_to_quantity
+)
+
+# Description
+# Amount of energy for body heat Q = C * m * (t2 - t1)
+# where:
+# C - specific heat capacity
+# m - body mass
+# t1 - temperature_origin (origin process)
```
Comment (blackyblack) [inline symplyphysics/definitions/amount_energy_from_mass_and_temperature.py:11]:
```suggestion
# t1 - initial temperature
```

File: symplyphysics/definitions/amount_energy_from_mass_and_temperature.py [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+# t2 - temperature_end ( end process)
+# Note: specific heat capacity C is parameter indicates how much energy in joules
+# must be expended to heat a unit mass of a substance by one unit temperature.
+# Note: the resultant energy is the energy consumed for changing body temperature from value t1 to value t2
+# Positive value Q is for absorbed energy( heating), - negative for released energy(cooling)
```
Comment (blackyblack) [inline symplyphysics/definitions/amount_energy_from_mass_and_temperature.py:16]:
```suggestion
# Positive value Q is for absorbed energy (heating), negative for released energy (cooling)
```

File: symplyphysics/definitions/power_is_proportional_voltage_and_current.py [hunk 27]
```diff
@@ -0,0 +1,23 @@
+from symplyphysics import (
+    symbols, Eq, pretty, solve, Quantity, units,
+    validate_input, validate_output, expr_to_quantity
+)
+
+# Description
+# Power of current is proportional to current and voltage
+# P = I * U
+# where:
+# I -the current flowing through element
```
Comment (blackyblack) [inline symplyphysics/definitions/power_is_proportional_voltage_and_current.py:10]:
```suggestion
# I - the current flowing through element
```

File: symplyphysics/definitions/power_is_proportional_voltage_and_current.py [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+# U - is voltage  on this element
```
Comment (blackyblack) [inline symplyphysics/definitions/power_is_proportional_voltage_and_current.py:11]: Add newline after description

File: symplyphysics/laws/electricity/amount_energy_from_voltage_time_resistance.py [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+# to the resistance of the conductor. This is the differential form of the law
+## This Joule-Lenz law is derived from more generic law:
+## See: [dissipated_heat_power](./dissipated_heat_power_is_proportional_to_current_square.py) implementation.
+
+# Amount of energy Q = U**2 * t / R
+# where:
+# U - voltage to conductor
+# t - time of current action
+# R - resistance of conductor
+# The resultant energy is the energy generated by the passage of current through the conductor.
```
Comment (blackyblack) [inline symplyphysics/laws/electricity/amount_energy_from_voltage_time_resistance.py:18]: Add newline

File: test/definitions/amount_energy_from_mass_and_temperature_test.py [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+# Density of water - 1000 kilogram / meter**3, 273 Kelvin == 0 Celsius_degree
+CELSIUS_TO_KELVIN_OFFSET = int(273)
+@fixture
+def test_args():
+    C = units.Quantity('C')
+    SI.set_quantity_dimension(C, units.energy / (units.mass * units.temperature))
+    SI.set_quantity_scale_factor(C, 4200 * units.joule / (units.kilogram * units.kelvin))
+    m = units.Quantity('m')
+    SI.set_quantity_dimension(m, units.mass)
+    SI.set_quantity_scale_factor(m, 0.5 * units.liter * 1000 * units.kilogram / units.meter**3)
```
Comment (blackyblack) [inline test/definitions/amount_energy_from_mass_and_temperature_test.py:20]: You can use kilograms here. Should be shorter. Do not forget to update description to use kilos as well.

File: test/definitions/amount_energy_from_mass_and_temperature_test.py [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+    t1 = units.Quantity('t1')
+    SI.set_quantity_dimension(t1, units.temperature)
+    SI.set_quantity_scale_factor(t1, (273 - CELSIUS_TO_KELVIN_OFFSET) * units.kelvin)
```
Comment (blackyblack) [inline test/definitions/amount_energy_from_mass_and_temperature_test.py:23]: Are you heating from 0 kelvin temperature? I suggest to drop celsius degrees here and use kelvin. Tests are kinda synthetic so you can precalc some values for shorter code.

File: symplyphysics/definitions/amount_energy_from_mass_and_temperature.py [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+# Amount of energy for body heat Q = C * m * (t2 - t1)
+# where:
+# C - specific heat capacity
+# m - body mass
+# t1 - initial temperature
+# t2 - final temperature
+# Note: specific heat capacity C is parameter indicates how much energy in joules
+# must be expended to heat a unit mass of a substance by one unit temperature.
+# Note: the resultant energy is the energy consumed for changing body temperature from value t1 to value t2
+# Positive value Q is for absorbed energy (heating), - negative for released energy (cooling)
```
Comment (blackyblack) [inline symplyphysics/definitions/amount_energy_from_mass_and_temperature.py:16]:
```suggestion
# Positive value Q is for absorbed energy (heating), negative for released energy (cooling)
```

**Review by blackyblack — APPROVED** [review]
